JS massage

frontierpsycho opened the pull request:
This is just for tracking, don't merge just yet.

frontierpsycho commented:
There's something odd going on with data initialization. I feel it can be done more neatly. Hm.

thiderman commented:
I've only skimmed through this, but it looks great! Keep the good stuff up! <3

thiderman commented:
You're committing with your work email, which is probably not your intention. You might want to [change the email address](http://stackoverflow.com/a/750182/983746).

frontierpsycho commented:
I'd say we can merge this now. Take a look and let me know. 

Some highlights:
* I made a `Tournament` and a `Match` object and used them widely
* I used some useful ES6 candy, and also `lodash` to avoid excessive looping (it is considered harmful)
* I fixed many (all?) the links to use named routes and params objects instead of string interpolation, which means links now work regardless of the presence of a final `/` in the URL
* I added tournament broadcasting in `App.vue`, so that `Tournament.vue` (and later, other views) can listen to changes to the tournament. In plain English, when people join from multiple places, ALL of them will see EVERYBODY else who's joining. Same thing for tournament progression: it will be shown to all connected to the WebSocket.
* Small things here and there, removed stuff we didn't need to use, etc.

frontierpsycho commented:
NOTE: Perhaps I could use the broadcast in Match.vue as well, which is doing it the old way right now. But we can still merge, and I can do that on master.

thiderman commented on js/src/components/Tournament.vue:
```diff
@@ -97,8 +97,6 @@ export default {
           let j = res.json()
           this.$route.router.go('/towerfall' + j.redirect)
-          // XXX: Worst hack of all time
-          this.$data.tournament.started = 'hehe'
```
Thanks for deleting this <3

thiderman commented on js/src/components/Tournament.vue:
```diff
@@ -70,94 +72,82 @@ export default {
-          }
-        }
+      // listen for tournaments from App
+      this.$on(`tournament${to.params.tournament}`, (tournament) => {
```
What's this syntax?

frontierpsycho replied:
It's the ES6 [string interpolation](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Template_literals) syntax. Basically, you surround your string with backticks, which allows you to evaluate expressions within ${}.

thiderman replied:
Ah, neat! The more you know! Thanks!

thiderman commented:
I've looked through all of this now! It looks great, and I am learning a lot! I really like the objectification of the API and the general separation of concerns. You also removed almost all of my `TODO: this is terrible :'(`, which feels super great!

10/10 feel free to merge! 👍

frontierpsycho commented:
Killing `// TODO blergh!` is my favourite coding hobby. I'm glad you like it! I'll keep at it once I find more tasks.
